src/aws/flb_s3_local_buffer.c

```diff
@@ -0,0 +1,384 @@
+    }
+
+  return -1;
+}
+
+/*
+ * "Initializes" the local buffer from the file system
+ * Reads buffer directory and finds any existing files
+ * This ensures the plugin will still send buffered data even if FB is restarted
+ */
+int flb_init_local_buffer(struct flb_local_buffer *store)
+{
+    DIR *d;
+    struct dirent *dir;
+    struct flb_local_chunk *c;
+    char *tag;
+    flb_sds_t path;
+    flb_sds_t tmp_sds;
+
+    d = opendir(store->dir);
+    if (d) {
+        while ((dir = readdir(d)) != NULL) {
+            if (dir->d_type == DT_REG) {
+                if (is_tag_file(dir->d_name) == 0) {
+                    continue;
+                }
+                /* create a new chunk */
+                flb_plg_debug(store->ins, "Found existing local buffer file %s",
+                              dir->d_name);
+                c = flb_calloc(1, sizeof(struct flb_local_chunk));
+                if (!c) {
+                    flb_errno();
+                    return -1;
+                }
+                c->create_time = time(NULL);
+                c->key = flb_sds_create(dir->d_name);
+                if (!c->key) {
+                    flb_errno();
+                    flb_chunk_destroy(c);
+                    return -1;
+                }
+                path = flb_sds_create_size(strlen(store->dir) + strlen(dir->d_name));
+                if (!path) {
+                    flb_errno();
+                    flb_chunk_destroy(c);
+                    flb_errno();
+                    return -1;
+                }
+                tmp_sds = flb_sds_printf(&path, "%s/%s", store->dir, dir->d_name);
+                if (!tmp_sds) {
+                    flb_errno();
+                    flb_chunk_destroy(c);
+                    flb_sds_destroy(path);
+                    return -1;
+                }
+                path = tmp_sds;
+                c->file_path = path;
+                /* get the fluent tag */
+                tag = read_tag(path);
+                if (!tag) {
+                    flb_plg_error(store->ins, "Could not read Fluent tag from file system; file path=%s.tag",
+                                  path);
+                    flb_errno();
+                    flb_chunk_destroy(c);
+                    return -1;
+                }
+                c->tag = flb_sds_create(tag);
+                flb_free(tag);
+                if (!c->tag) {
+                    flb_errno();
+                    flb_chunk_destroy(c);
+                    return -1;
+                }
+                flb_plg_info(store->ins, "Found existing local buffer %s",
+                             path);
+                mk_list_add(&c->_head, &store->chunks);
+            }
+        }
+        closedir(d);
+    }
+  return 0;
+}
+
+/*
+ * Recursively creates directories
+ */
+int flb_mkdir_all(const char *dir) {
+    char tmp[PATH_MAX];
+    char *p = NULL;
+    int ret;
+    size_t len;
+
+    snprintf(tmp, sizeof(tmp), "%s", dir);
```

Comment:
Why did you switch from the SDS string functions to standard library functions? Should you be using `flb_sds_printf()` here?